extensions/amp-video/0.1/video-cache.js

```diff
@@ -215,3 +198,58 @@ function getCacheUrlService(videoEl, ampdoc) {
     .installExtensionForDoc(ampdoc, 'amp-cache-url')
     .then(() => Services.cacheUrlServicePromiseForDoc(videoEl));
 }
+
+/**
+ * Fetch the sources for the given video element.
+ * @param {!Element} videoEl
+ * @param {!AmpDoc} ampdoc
+ * @return {!Promise<!Object>} JSON representing AMP's cached video sources.
+ */
+function requestCachedVideoSources(videoEl, ampdoc) {
+  const {win} = ampdoc;
+  if (shouldUseInlineVideoResponse(videoEl)) {
+    const inlineResponseEl = win.document.getElementById(
+      'amp-google-video-cache-response'
+    );
+    try {
+      const inlineResponseJson = JSON.parse(inlineResponseEl.textContent);
+      if (inlineResponseJson['sources']) {
+        return Promise.resolve(inlineResponseJson);
+      }
+    } catch (err) {
+      // If parsing the response fails, an XHR request will be made below.
+    }
+  }
+
+  const {canonicalUrl, sourceUrl} = Services.documentInfoForDoc(win.document);
+  maybeReplaceSrcWithSourceElement(videoEl, win);
+  const videoUrl = resolveRelativeUrl(selectVideoSource(videoEl), sourceUrl);
+  return getCacheUrlService(videoEl, ampdoc)
+    .then((service) => service.createCacheUrl(videoUrl))
+    .then((cacheUrl) => {
+      const requestUrl = addParamsToUrl(cacheUrl.replace(/\/[ic]\//, '/mbv/'), {
+        'amp_video_host_url':
+          /* document url that contains the video */ canonicalUrl,
+        'amp_video_require_acao_header': videoEl.hasAttribute('crossorigin')
+          ? 1
+          : null,
+      });
+      return Services.xhrFor(win)
+        .fetch(requestUrl, {prerenderSafe: true})
+        .then((xhrResponse) => xhrResponse.json());
+    });
+}
+
+/**
+ * Returns `true` if the video's inline response should be used instead of
+ * issuing an XHR request.
+ * @param {!Element} videoEl
+ * @return {boolean}
+ */
+function shouldUseInlineVideoResponse(videoEl) {
+  // Google video cache inlines the first video of the first web story page.
+  return matches(
+    videoEl,
+    `amp-story-page:first-of-type amp-video:first-of-type`
+  );
```

Comment:
You were right that this could potentially match multiple amp-video elements if they are not siblings.

Should we switch to something like `win.document.querySelector('amp-story-page:first-of-type amp-video') === videoEl`?